src/bigtable/mod.rs

```diff
@@ -0,0 +1,537 @@
+            .families
+            .last_mut()
+            .and_then(|family| family.columns.last_mut())
+    }
+
+    fn new_cell(&mut self, timestamp_micros: i64, value_size: i32) {
+        self.flush_bytes_in_progress();
+        if let Some(col) = self.last_column_mut() {
+            col.cells.push(Cell {
+                timestamp_micros,
+                ..Default::default()
+            });
+            self.value = BytesMut::with_capacity(value_size as usize);
+        }
+    }
+
+    fn last_cell_mut(&mut self) -> Option<&mut Cell> {
+        self.last_column_mut().and_then(|col| col.cells.last_mut())
+    }
+
+    fn finish_row(&mut self) -> Row {
+        self.flush_bytes_in_progress();
+        std::mem::replace(&mut self.row, Default::default())
+    }
+}
+
+/// A stream of rows, returned in response to a `read_rows` request.
+pub struct RowStream {
+    // TODO: once retries are handled correctly, this will get replaced by
+    // something else. Probably a manual implementation of Stream.
+    inner: Pin<Box<dyn futures::Stream<Item = Result<Row, tonic::Status>>>>,
+}
+
+impl futures::Stream for RowStream {
+    type Item = Result<Row, tonic::Status>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut futures::task::Context<'_>,
+    ) -> futures::task::Poll<Option<Self::Item>> {
+        self.inner.poll_next_unpin(cx)
+    }
+}
+
+/// A client for connecting to bigtable. Created from the
+/// [`build_bigtable_client`](crate::builder::ClientBuilder::build_bigtable_client)
+/// function.
+#[derive(Clone)]
+pub struct BigtableClient<
+    C = crate::DefaultConnector,
+    Retry = ExponentialBackoff<BigtableRetryCheck>,
+> {
+    inner: api::bigtable::v2::bigtable_client::BigtableClient<
+        AuthGrpcService<tonic::transport::Channel, OAuthTokenSource<C>>,
+    >,
+    retry: Retry,
+    table_prefix: String,
+}
+
+impl<C, Retry> BigtableClient<C, Retry>
+where
+    C: crate::Connect + Clone + Send + Sync + 'static,
+    Retry: RetryPolicy<(), tonic::Status> + 'static,
+    Retry::RetryOp: Send + 'static,
+    <Retry::RetryOp as RetryOperation<(), tonic::Status>>::Sleep: Send + 'static,
+{
+    /// Request some rows from bigtable.
+    ///
+    /// This is the most general read request; various other convenience methods are
+    /// available.
+    pub async fn read_rows(
+        &mut self,
+        request: ReadRowsRequest,
+    ) -> Result<RowStream, tonic::Status> {
+        let response = self.inner.read_rows(request).await?.into_inner();
+
+        // Turn the response into a stream of rows.
+        // TODO: on errors, we should follow the retry policy. This is a little complicated, because
+        // on error we need to re-issue the query, while taking into account the last_scanned_row_key field
+        // to reduce re-querying. See the go implementation at
+        // https://github.com/googleapis/google-cloud-go/blob/21fc86d799bcdffc3db6f697b30ca48e072e7044/bigtable/bigtable.go#L171
```

Comment:
I'm not sure if I understood this correctly, but wouldn't non-smart re-queries result in the user seeing repeated rows in the stream? Anyway, it might be moot because the updated PR attempts the correct retry behavior.

Reply:
Ah you're right, I hadn't considered elements already yielded to the caller. Anyway, yeah real retries are better